src/irmin/proof.ml

```diff
@@ -132,9 +142,57 @@ struct
   let empty () : t = ref Empty
   let is_empty t = !t = Empty
 
+  type hash = H.t [@@deriving irmin ~equal ~pp]
+
+  let rec forward_lookup h singleton_inodes : (int list * hash) list option =
+    match Hashes.find_opt singleton_inodes h with
+    | None -> None
+    | Some (i, h') -> (
+        match forward_lookup h' singleton_inodes with
+        | None -> Some [ (i, h') ]
+        | Some l -> Some ((i, h') :: l))
+
+  let rec extenders (indexes, h) = function
+    | (i', h') :: rest -> extenders (i' @ indexes, h') rest
```

Comment:
`List.rev i'` ? the backend currently only returns singletons, but that should also work for incomplete extenders.